plugins/gradle/src/org/jetbrains/plugins/gradle/service/settings/GradleDaemonJvmCriteriaView.kt

```diff
@@ -0,0 +1,122 @@
+import com.intellij.ui.dsl.builder.*
+import com.intellij.ui.dsl.listCellRenderer.textListCellRenderer
+import com.intellij.util.ui.UIUtil
+import org.jetbrains.annotations.VisibleForTesting
+import org.jetbrains.plugins.gradle.properties.models.Property
+import org.jetbrains.plugins.gradle.util.GradleBundle
+import javax.swing.JPanel
+import javax.swing.event.DocumentEvent
+import javax.swing.plaf.basic.BasicComboBoxEditor
+
+private const val UNDEFINED_VERSION_PLACEHOLDER = "UNDEFINED"
+private const val ANY_VENDOR_PLACEHOLDER = "<ANY_VENDOR>"
+
+class GradleDaemonJvmCriteriaView(
+  version: Property<String>?,
+  vendor: Property<String>?,
+  private val versionsDropdownList: IntRange,
+  private val vendorDropdownList: List<String>,
+  private val displayAdvancedSettings: Boolean
+): JPanel(VerticalLayout(0)) {
+
+  val selectedVersion: String
+    get() = versionComboBox.editor.item.toString()
+  val selectedVendor: String?
+    get() = vendorComboBox.editor.item?.toString()?.takeIf { it.isNotBlank() || it != ANY_VENDOR_PLACEHOLDER }
+  val isModified: Boolean
+    get() = selectedVersion != initialVersion || selectedVendor != initialVendor
+  val isValidVersion: Boolean
+    get() = selectedVersion.toIntOrNull() != null
+  val isValidVendor: Boolean
+    get() = selectedVendor?.isNotBlank() == true
+
+  private var initialVersion: String = version?.value ?: UNDEFINED_VERSION_PLACEHOLDER
+  private var initialVendor: String = vendor?.value ?: ANY_VENDOR_PLACEHOLDER
+
+  @get:VisibleForTesting
+  lateinit var versionComboBox: ComboBox<Int>
+  @get:VisibleForTesting
+  lateinit var vendorComboBox: ComboBox<String>
+
+  init {
+    val gradleJvmView = panel {
+      row {
+        comment(GradleBundle.message("gradle.settings.text.daemon.toolchain.title")).applyToComponent {
+          font = UIUtil.getLabelFont(UIUtil.FontSize.SMALL)
+          foreground = UIUtil.getLabelFontColor(UIUtil.FontColor.BRIGHTER)
+        }
+      }
+      row {
+        label(GradleBundle.message("gradle.settings.text.daemon.toolchain.version"))
+        versionComboBox = comboBox(versionsDropdownList.toList().reversed(), textListCellRenderer { it.toString() })
+          .columns(COLUMNS_SHORT)
+          .applyToComponent {
+            isEditable = true
+            isFocusable = false
+            editor = GradleJdkPathComboBoxEditor(this) { toIntOrNull() != null }
+            editor.item = initialVersion
+          }.component
+      }
+      collapsibleGroup(ApplicationBundle.message("title.advanced.settings")) {
+        row {
+          label(GradleBundle.message("gradle.settings.text.daemon.toolchain.vendor"))
+          vendorComboBox = comboBox(listOf(ANY_VENDOR_PLACEHOLDER) + vendorDropdownList, textListCellRenderer { it })
+            .columns(COLUMNS_SHORT)
+            .applyToComponent {
+              isEditable = true
+              editor = GradleJdkPathComboBoxEditor(this) { isNotBlank() }
```

Comment:
Why do you need this custom editor? This isn't a straight way to define validation in Kotlin UI DSL. Also, your approach doesn't work for me.

<img width="1002" alt="image" src="https://github.com/user-attachments/assets/d6817d76-e666-4901-a65f-57a3ed7fc43d">

Reply:
You can check this example if you need this validation
```
    val gradleJvmView = panel {
      row {
        comment(GradleBundle.message("gradle.settings.text.daemon.toolchain.title")).applyToComponent {
          font = UIUtil.getLabelFont(UIUtil.FontSize.SMALL)
          foreground = UIUtil.getLabelFontColor(UIUtil.FontColor.BRIGHTER)
        }
      }
      row {
        label(GradleBundle.message("gradle.settings.text.daemon.toolchain.version"))
        versionComboBox = comboBox(versionsDropdownList.toList().reversed(), textListCellRenderer { it.toString() })
          .columns(COLUMNS_SHORT)
          .applyToComponent {
            selectedItem = initialVersion
          }.component
      }
      collapsibleGroup(ApplicationBundle.message("title.advanced.settings")) {
        row {
          label(GradleBundle.message("gradle.settings.text.daemon.toolchain.vendor"))
          vendorComboBox = comboBox(listOf(ANY_VENDOR_PLACEHOLDER) + vendorDropdownList, textListCellRenderer { it })
            .columns(COLUMNS_SHORT)
            .validationOnInput {
              if (!isValidVendor) error("Invalid vendor") else null
            }
            .applyToComponent {
              isEditable = true
              selectedItem = initialVersion
            }.component
        }
        row {
          comment(GradleBundle.message("gradle.settings.text.daemon.toolchain.vendor.hint")).applyToComponent {
            font = UIUtil.getLabelFont(UIUtil.FontSize.SMALL)
            foreground = UIUtil.getLabelFontColor(UIUtil.FontColor.BRIGHTER)
          }
        }
      }.topGap(TopGap.NONE)
        .visible(displayAdvancedSettings)
    }.apply {
      registerValidators(parentDisposable)
    }
```

<img width="1322" alt="image" src="https://github.com/user-attachments/assets/020997dd-cae3-4263-aa8e-e8f2745b4b5c">

Reply:
The `GradleJdkPathComboBoxEditor ` was added to validate input specially for version then non numeric value is specified. However, as suggested I used `validationOnInput` since already provides the required API to archive the same much simpler :)